viewing_party/party.py

```diff
@@ -1,23 +1,201 @@
+def get_friends_unique_watched(user_data):
+    # print(f"user_data[watch]: {user_data['watched']}")
+    # print("******************************************")
+    # print(f"user_data[firends]: {user_data['friends']}")
+    # print("******************************************")
+    #user watched set
+    user_watched_set = set()
+    for movie in user_data["watched"]:
+        user_watched_set.add(movie["title"])
+    # friends watch set
+    friends_watched_set = set()
+    for friend_watched in user_data["friends"]:
+        for movie_name in friend_watched["watched"]:
+            friends_watched_set.add(movie_name["title"])
+    #friends watched but user's didn't
+    user_nonwatched_set = friends_watched_set.difference(user_watched_set)
+    # print(f"user_nonwatched_set: {user_nonwatched_set}")
+    # print("******************************************")
+    # create the result list to store
+    # use title value to find the movie dictionaries and store them to the result list
+    result_list =[]
+    for name in list(user_nonwatched_set):
+        for friends_movie in user_data["friends"]:
+            for movie in friends_movie["watched"]:
+                if name == movie["title"] and result_list.count(movie)==0 :
+                    result_list.append(movie)
+    # print(f"result_list: {result_list}")
+    return result_list
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    # return a recommended movies list
+    # conditions:
+    # user didn't watch but friend watched
+    # match the user's subscriptions list
+    # may use  get_friends_unique_watched(user_data) to return a user non_watched list
+    user_nonwatched_list = get_friends_unique_watched(user_data)
+    user_subscriptions_list = user_data["subscriptions"]
+    recommended_movies_list = []
+    for movie in user_nonwatched_list:
+        for host_name in user_subscriptions_list:
+            if movie["host"] == host_name:
+                recommended_movies_list.append(movie)
+    return recommended_movies_list
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    # return a recommended movies list
+    # conditions:
+    # user didn't watch but friends watched
+    # The "genre" match the user's most frequent genre
+    # may use get_most_watched_genre() to get the most frequent genre
+    # and use get_friends_unique_watched() to get a user non_watched list
+    max_genre = get_most_watched_genre(user_data)
+    user_nonwatched_list = get_friends_unique_watched(user_data)
+    recommended_movies_list = []
+    for movie in user_nonwatched_list:
+        if movie["genre"] == max_genre:
+            recommended_movies_list.append(movie)
+    return recommended_movies_list 
+
+def get_rec_from_favorites(user_data):

```

Comment:
This implementation gets us what we need, but I think we could benefit by asking questions to see what we could simplify. Some to start with could be:
- There are several functions in this file that take a list of movies create a set of just the titles. What helper functions could we create to reduce duplication of the set creation code? 
- Are there functions we already wrote that could be used to help solve this problem?